README.md
[![get-latest-offers](https://github.com/Ffyud/kortings/actions/workflows/get-latest-offers.yml/badge.svg)](https://github.com/Ffyud/kortings/actions/workflows/get-latest-offers.yml) [![deploy-to-gh-pages](https://github.com/Ffyud/kortings/actions/workflows/deploy-to-gh-pages.yml/badge.svg)](https://github.com/Ffyud/kortings/actions/workflows/deploy-to-gh-pages.yml) [![run-parse-offers-test](https://github.com/Ffyud/kortings/actions/workflows/run-parse-offers-test.yml/badge.svg)](https://github.com/Ffyud/kortings/actions/workflows/run-parse-offers-test.yml) [![code-coverage](./parse_offers/coverage-badge.svg?dummy=8484744)](./parse_offers/htmlcov/index.html)

# Teerkost
![](https://teerkost.nl/logo192.png)

[https://teerkost.nl](https://teerkost.nl)

Teerkost collects the current discounts from popular Dutch supermarkets.

- It is fast - there is no backend it needs to call upon.
- It is uncluttered - filter by category, shop, discount or do a search.
- Bookmark offers by saving them locally using IndexedDb (i.e. "cookies").
- It is made to feel like an app.

## Running or building the app

Start the app called *kortings-app* locally:
```
npm start
```

## Getting discounts

The data is collected trough a Python package called *parse_offers*. 

It will populate a json file containing all the discounts and move it to the app source code.

Run the Python script:

```
python parse_offers
```

Run the tests from within the parse_offers folder:
```
python -m unittest -b -v
```
Run the tests and output Junit XML reporting:
```
python -m xmlrunner discover
```
Generate a code coverage report
```
coverage run -m unittest
coverage report
coverage html
```


## Continuous deployment

Through Github workflows the app is continously being updated, tested and deployed to Github Pages.

### 1. run-parse-offers-test.yml
Runs unittests against the parse_offers package to see if it returns complete results.

### 2. run-py-test-coverage.yml
Generates a code coverage report for the parse_offers package.

### 3. get-latest-offers.yml
Runs parse_offers by a daily schedule. A json file is populated with the offers found.

### 4. deploy-to-gh-pages.yml
Creates a build from the *kortings-app* and pushes it to the *gh-pages* branch. Github Pages uses this branch as the source.

### 5 Pylint on CircleCi
CircleCi [runs a code quality review](https://app.circleci.com/pipelines/github/Ffyud/teerkost?branch=main) of parse_offers with *Pylint*.

CHANGELOG.md
# Changelog

All notable changes to the project are documented here. There is no versioning for now. Just what is changed on what date.

# 18-10-2022

### Added
- Buttons for filtering by discount percentages.

# 17-10-2022

### Added
- CI pipeline: Code quality scanning with Pylint on every change.
- CI pipeline: Code coverage scanning on every change, including SVG badge in readme.

### Changed
- Refactor in parse_offers related to getting the right date (new module).
- Refactor in parse_offers related to price and deals (new module).

### Fixed
- Fixed many code quality issues that came up with Pylint.

# 07-10-2022

### Added
- Improved categorizing by text with multiple words and whitespace.
  
# 06-10-2022

### Changed
- Skip collecting the huge amount of non-food offers from 'Aldi'.

# 01-10-2022

### Fixed
- Fix for resetting scroll to top position when changing a filter.

# 30-09-2022

### Added
- Category filter for 'siroop'.

### Changed
- Combined two categories into 'zoetigheid'.
  
# 23-09-2022

### Fixed
- Fix for incomplete url when sharing a bookmarks list.
  
### Added
- Category filter for 'dier'.
- Category filter for 'sap'.
- Category filter for uncategorized offers.

# 10-09-2022

### Changed
- Deal text more consistent across shops.

# 04-09-2022

### Fixed
- Fix for wrong deal text from 'Spar'.
- Fix for no date or deal found from 'Lidl'.

# 23-07-2022

### Added
- Shop 'Spar' to the parser.
  
# 22-07-2022

### Added
- Shop 'Dirk' to the parser.

# 18-07-2022

### Added
- Clickable title to offers, to visit the shop homepage.
- Actual prices to offers from 'Lidl'.

### Changed
- Saving offers no longer disrupted when a shop is unreachable.

### Fixed
- Fix for not collecting data from 'Lidl'.
- For for unintentional clicks on donate button.

# 25-06-2022

### Added
- Ability to search for offers.

# 18-06-2022

### Fixed
- Fix for properly filling display width with items.

# 23-05-2022

### Fixed
- Fix for not collecting offers from 'Lidl' when only startdate is provided.
 
# 20-05-2022

### Changed
- Style of subtitle on page to be uniform across screen sizes.
  
# 18-05-2022

### Added
- Explainer text to no-bookmarks, nothing-found and 404 pages.
- Navigation bar and title to the 404 page.

### Fixed
- Fix when price sometimes is not shown for 'Plus'.

# 17-05-2022

### Added
- Junit style test reporting to be used by Github actions.

# 11-05-2022

### Changed
- App icons.

# 06-05-2022

### Changed
- Bottom navigation component on small screens.
- Home button on the bookmarks screen.

### Fixed
- Fix for wrong page title on other pages than home.
- Fix bottom padding for non‑rectangular displays.

# 02-05-2022

### Added
- Shop 'Plus' to the parser.

# 01-05-2022

### Changed
- Small color changes to the theme.
- Two column layout on small screens.

# 29-04-2022

### Added
- Displaying offers as list by id's, so the user can share bookmarks.

# 28-04-2022

### Changed
- Clean up more redundant text from product info.
- Small styling changes for the item cards.

### Fixed
- Fix for removing offer bookmarks that do not exist anymore.
  
# 26-04-2022

### Changed
- Item cards have a cleaner look.

### Fixed
- Fix for overflowing date label when item gets to small.

# 25-04-2022

### Added
- Copy the bookmarked offer as a shopping list.

# 24-04-2022

### Changed
- Wider sharing pop-up and added some information.

# 22-04-2022

### Changed
- Filters stay sticky when screenheight is big enough.

### Fixed
- Fix alignment of page on very large screens.
- Fix showing an empty filter without a category.

# 21-04-2022

### Added
- Category filter for 'gebak'.
- Category filter for 'tuin'.
- Visual feedback to copy-link when sharing the page.

### Fixed
- Fix for not finding offers from 'Aldi' when startday is not monday.

# 20-04-2022

### Added
- Clickable 'Teerkost' title added to the small screen version.

### Changed
- Sort offers according to the ordering of the filter labels.
- Refactor categories to seperate file.
- Clickable 'Teerkost' title always aligned to center of page.

# 19-04-2022

### Added
- Category filter for 'snoep'.

### Fixed
- Fix for not up-to-date state of bookmarked item when navigating.

### Changed
- Cleans up the offer info text a little better.

# 18-04-2022

### Added
- First version of bookmarking seperate offers, saving them locally in the browser.

### Fixed
- Fix for wrong validity period of 'AH' offers.
- Fix double "korting" label in some offers.

### Changed
- Size of date label on small screens.
- Show date instead of 'verlopen'.
- Removes redundant logo on share-dialog.
- Temporarily removes overlay with linking options on individual offers.
  
# 14-04-2022
### Fixed
- Fix for when 'Jumbo' does not provide an image.

### Added
- Unittests for collecting 'Ekoplaza' offers.

### Changed
- Provide a consistent product id to every offer item.
- Remove link to source code.
  
# 13-04-2022
### Added
- Category filter for 'ontbijt'.
- 'Load more' button for lazy loading and increased performance.
  
### Changed
- Sort filters buttons according to most used/important.

### Fixed
- Fix setting the wrong category when to-be-ignored word is found first.

# 12-04-2022
### Added
- Shop 'Ekoplaza' to the parser.
- Category filter for 'thee'.

# 11-04-2022
### Added
- Proper intention links for social media sharing.
- Date label to included 'verlopen' when offer is not current anymore.

# 10-04-2022
### Added
- Sharing dialog for browsers that do not support native sharing

# 08-04-2022
### Added
- Donation button
- Open Graph Protocol meta tags for rich social media sharing
- Update URL and page title according to the filters set

### Changed
- Move filters to the left when on a big enough screen

# 07-04-2022
### Added
- 404 'not-found' fallback page.

### Changed
- Design for displaying none-found and loading.

# 02-04-2022
### Added
- Numbering (amount of offers) to filter buttons.
- Words-to-be-ignored when categorizing products.

parse_offers/spar.py
from xml.etree.ElementTree import ElementTree
import requests
from cleanup import categorize, cleantext, cleandeal
from datetime import datetime
from bs4 import BeautifulSoup

def return_offers():

    SHOP = "spar"
    URL = "https://www.spar.nl/aanbiedingen/"

    r = requests.get(URL)
    soup = BeautifulSoup(r.content, "html.parser")

    element_with_dates = soup.select_one(".c-featured-promo__content")
    if element_with_dates is not None:
        content = element_with_dates.find("div", {"class":"content"}).get_text().strip()
        if "de aanbiedingen uit deze folder zijn geldig van" in content:
            splitcontent = content.split(" ")
            # de aanbiedingen uit deze folder zijn geldig van 21-07-2022 t/m 03-08-2022
            splitcontent.reverse()
            # 03-08-2022 t/m 21-07-2022 van geldig zijn folder deze uit aanbiedingen de
            date_start = splitcontent[2]
            date_start_string = datetime.strptime(date_start, "%d-%m-%Y").date()
            date_end = splitcontent[0]
            date_end_string = datetime.strptime(date_end, "%d-%m-%Y").date()
        else:
            print("Kan geen info vinden over start- en einddatum!")


    collection = []
    for item in soup.find_all("div", {"class": "c-product-tile"}):
        offer = {
            "productId": "",
            "product":"", 
            "productInfo":"", 
            "category":"", 
            "image":"", 
            "deal":"",
            "price": float(0), 
            "percentage":0,
            "dateStart":"", 
            "dateEnd":"", 
            "link": "", 
            "shop":""
        }

        clean_title = ""
        clean_info = ""

        meta_element = item.find("div", {"class":"c-product-tile__meta"})
        title_element = meta_element.find("a")
        if title_element is not None:
            title = title_element.get_text().strip()
            clean_title = cleantext.clean_up_title(title)

        description_element = meta_element.find("span")
        if description_element is not None:
            description = description_element.get_text().strip()
            clean_info = cleantext.clean_up_info(description)

        category = categorize.find_category(clean_title, clean_info)

        image_tile = item.find("div", {"class":"c-product-tile__image"})
        image_src = image_tile.find("img")
        image_url = ""
        if image_src is not None:
            image_url = image_src['data-src']

        full_href = ""
        product_id = ""

        link_element = image_tile.find("a")
        if link_element is not None:
            link = link_element['href']
            # https://www.spar.nl/aanbiedingen/kip-aanbieding-6861/
            full_href = "https://spar.nl" + link

            id_from_link = link.split("/")
            id_from_link.reverse()
            product_id = id_from_link[1]

        price_element = item.find("div", {"class":"c-product-tile__image--pricing"})
        if price_element is not None:
            old_price = price_element.find("span", {"class":"c-price__old"})
            if old_price is not None:
                new_price = price_element.find("span", {"class":"is-danger"})
                if new_price is not None:
                    base = new_price.find("span", {"class":"c-price__base"}).get_text().strip()
                    decimal = new_price.find("span", {"class":"c-price__mod"}).get_text().strip()
                    new_price = base + decimal
                    new_price = new_price.rstrip(".-")
                    old_price = old_price.get_text().strip().rstrip(".-")
                    offer.update({"price": float(new_price)})

                    calculate_deal = cleandeal.calculate_percentage(old_price, new_price)
                    percentage = int(float(calculate_deal))
                    deal = calculate_deal + "% korting"
                else:
                    print("Geen nieuwe prijs, geen deal")
            else:
                print("geen oude prijs, geen deal")
        else:
            # Geen prijs te zien, dus pagina van aanbieding zelf openen
            # FIXME prijs berekenen op basis van promotion tekst
            if full_href != "":
                r = requests.get(full_href)
                soup = BeautifulSoup(r.content, "html.parser")
                promotion_summary = soup.find("div", {"class":"c-promotion-summary__content"})
                promotion_summar_deal = promotion_summary.select_one("h4")
                if promotion_summar_deal is not None:
                    promotion = promotion_summar_deal.get_text().strip()
                    if len(promotion) > 20:
                        if 'nu met' in deal:
                            deal = deal.replace('nu met', '')
                        else:
                            if 'voor' in deal:
                                promotion_splitted = promotion.split("voor")
                                promotion_splitted.reverse()
                                deal = "voor" + promotion_splitted[0]
                    else:
                        deal = promotion
                else:
                    print("Kan geen deal vinden op " + full_href)
            else:
                print("Geen prijs en geen URL")

        deal = deal.replace('!', '')
        deal = deal.replace('Nu ', '')
        deal = deal.replace('nu ', '')
        deal = deal.replace('stuks voor', 'voor')

        offer.update({"productId": product_id})
        offer.update({"product": clean_title})
        offer.update({"productInfo": clean_info})
        offer.update({"category": category})
        offer.update({"link": full_href})
        offer.update({"deal": deal })
        offer.update({"percentage": percentage})
        offer.update({"image": image_url})
        offer.update({"dateStart": str(date_start_string)})
        offer.update({"dateEnd": str(date_end_string)})
        offer.update({"shop": SHOP})

        if offer.get('deal') != "": # if no deal is found, don't add it
            collection.append(offer)

    print("📄 " + str(len(collection)) + " aanbiedingen van de "+SHOP+" bij elkaar verzameld.")

    return collection

if __name__ == "__main__":
    return_offers()